### PR TITLE
fix: add gst_tax_type field

### PR DESCRIPTION
**Pull request by Sanket322**
Added `gst_tax_type` field in taxes table to have record of tax account

### CODE REVIEW

File: india_compliance/patches/v14/set_gst_tax_type.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    parent_doctype = frappe.qb.DocType(doctype, alias="parent_doctype")
+    child_doctype = frappe.qb.DocType(child_doctype, alias="child_doctype")
+
+    (
+        frappe.qb.update(child_doctype)
+        .join(parent_doctype)
+        .on(child_doctype.parent == parent_doctype.name)
+        .set(
+            child_doctype.gst_tax_type,
+            Case()
```
Comment (ljain112) [inline india_compliance/patches/v14/set_gst_tax_type.py]: loop

File: india_compliance/gst_india/constants/__init__.py [hunk 2]
```diff
@@ -13,6 +13,13 @@
     "cess_account",
     "cess_non_advol_account",
 )
+GST_REVERSE_CHARGE = (
```
Comment (ljain112) [inline india_compliance/gst_india/constants/__init__.py]: loop with gst tax types

File: india_compliance/gst_india/doctype/purchase_reconciliation_tool/__init__.py [hunk 3]
```diff
@@ -495,7 +489,7 @@ def query_tax_amount(self, account):
             Sum(
                 Case()
                 .when(
-                    self.PI_TAX.account_head == account,
+                    self.PI_TAX.gst_tax_type == account,
```
Comment (ljain112) [inline india_compliance/gst_india/doctype/purchase_reconciliation_tool/__init__.py]: account ??

File: india_compliance/gst_india/doctype/purchase_reconciliation_tool/__init__.py [hunk 4]
```diff
@@ -643,7 +635,7 @@ def query_tax_amount(self, account):
             Sum(
                 Case()
                 .when(
-                    self.BOE_TAX.account_head == account,
+                    self.BOE_TAX.gst_tax_type == account,
```
Comment (ljain112) [inline india_compliance/gst_india/doctype/purchase_reconciliation_tool/__init__.py]: .

File: india_compliance/gst_india/overrides/transaction.py [hunk 5]
```diff
@@ -306,24 +307,21 @@ def validate_gst_accounts(doc, is_sales_transaction=False):
...
-
-    def _get_matched_idx(rows_to_search, account_head_list):
+    def _get_matched_idx(rows_to_search, account_list):
         return next(
-            (
-                row.idx
-                for row in rows_to_search
-                if row.account_head in account_head_list
-            ),
+            (row.idx for row in rows_to_search if row.gst_tax_type in account_list),
```
Comment (ljain112) [inline india_compliance/gst_india/overrides/transaction.py]: .

File: india_compliance/gst_india/overrides/transaction.py [hunk 6]
```diff
@@ -335,9 +333,6 @@ def _throw(message, title=None):
         for_sales=is_sales_transaction,
```
Comment (ljain112) [inline india_compliance/gst_india/overrides/transaction.py]: .

File: india_compliance/gst_india/overrides/transaction.py [hunk 7]
```diff
@@ -491,10 +480,52 @@ def _throw(message, title=None):
...
+
+        for account, tax_name in gst_accounts.items():
+            account = (
+                account[:-8]
+                if account_type != "Reverse Charge"
+                else account[:-8] + "_rcm"
+            )
+
+            if tax_name in used_accounts:
+                tax_name_to_tax_type[tax_name] = account
```
Comment (ljain112) [inline india_compliance/gst_india/overrides/transaction.py]: variable naming

File: india_compliance/gst_india/report/gstr_3b_details/gstr_3b_details.py [hunk 8]
```diff
@@ -494,7 +495,7 @@ def get_tax_case_statement(account, alias):
             return Sum(
                 Case()
                 .when(
-                    taxes.account_head.isin(account),
+                    taxes.gst_tax_type.isin(account),
```
Comment (ljain112) [inline india_compliance/gst_india/report/gstr_3b_details/gstr_3b_details.py]: .

File: india_compliance/patches.txt [hunk 9]
```diff
@@ -51,4 +51,5 @@ india_compliance.patches.v14.update_default_gstr1_settings
 india_compliance.patches.v14.add_match_found_in_purchase_reconciliation_status
 india_compliance.patches.v14.unset_inward_supply_link_for_cancelled_purchase
```
Comment (ljain112) [inline india_compliance/patches.txt:52]: custom field

File: india_compliance/gst_india/constants/__init__.py [hunk 10]
```diff
@@ -13,16 +13,11 @@
...
-    "cgst_rcm",
-    "sgst_rcm",
-    "igst_rcm",
-    "cess_rcm",
-    "cess_non_advol_rcm",
-)
 
 GST_TAX_TYPES = tuple(field[:-8] for field in GST_ACCOUNT_FIELDS)
 
+GST_REVERSE_CHARGE = tuple(field + "_rcm" for field in GST_TAX_TYPES)
```
Comment (ljain112) [inline india_compliance/gst_india/constants/__init__.py]: tax type

File: india_compliance/gst_india/overrides/transaction.py [hunk 11]
```diff
@@ -488,13 +512,47 @@ def _throw(message, title=None):
...
-    if (
-        tax_row.charge_type == "On Item Quantity"
-        and tax_row.account_head not in cess_non_advol_accounts
+    gst_tax_account_map = {}
+    is_sales_transaction = (
+        True if doc.doctype in (*SALES_DOCTYPES, "Payment Entry") else False
+    )
+    account_types = ["Output"] if is_sales_transaction else ["Input", "Reverse Charge"]
+
+    used_accounts = [
```
Comment (vorasmit) [inline india_compliance/gst_india/overrides/transaction.py]: utility for gst account type

File: india_compliance/gst_india/overrides/transaction.py [hunk 12]
```diff
@@ -408,35 +467,9 @@ def _throw(message, title=None):
                 ).format(idx)
             )
 
-    is_inter_state = is_inter_state_supply(doc)
     previous_row_references = set()
 
     for row in rows_to_validate:
-        account_head = row.account_head
```
Comment (vorasmit) [inline india_compliance/gst_india/overrides/transaction.py:415]: revert validation

File: india_compliance/gst_india/overrides/ineligible_itc.py [hunk 13]
```diff
@@ -272,11 +272,10 @@ def update_ineligible_taxes(self, item):
             "Input SGST - FC": 50,
         }
         """
-        gst_accounts = get_gst_accounts_by_type(self.doc.company, "Input").values()
         ineligible_taxes = frappe._dict()
 
         for tax in self.doc.taxes:
-            if tax.account_head not in gst_accounts:
+            if tax.gst_tax_type not in GST_TAX_TYPES:
```
Comment (vorasmit) [inline india_compliance/gst_india/overrides/ineligible_itc.py:278]: use TaxTypes at all places

File: india_compliance/gst_india/utils/tests.py [hunk 14]
```diff
@@ -146,6 +146,7 @@ def _append_taxes(
             "description": account,
             "rate": rate,
             "cost_center": f"Main - {company_abbr}",
+            "gst_tax_type": account.replace(" ", "_").lower(),
```
Comment (vorasmit) [inline india_compliance/gst_india/utils/tests.py]: why is it not set from transaction before validate?

**Pull request merged**